src/v/cluster/rm_stm.cc

```diff
@@ -1719,11 +1742,19 @@ void rm_stm::apply_fence(model::record_batch&& b) {
     iobuf_parser val_reader(std::move(val_buf));
     auto version = reflection::adl<int8_t>{}.from(val_reader);
     vassert(
-      version == rm_stm::fence_control_record_version,
+      version <= rm_stm::fence_control_record_version,
       "unknown fence record version: {} expected: {}",
       version,
       rm_stm::fence_control_record_version);
 
+    std::optional<model::tx_seq> tx_seq{};
+    std::optional<std::chrono::milliseconds> transaction_timeout_ms;
+    if (version == rm_stm::fence_control_record_version) {
+        tx_seq = reflection::adl<model::tx_seq>{}.from(val_reader);
+        transaction_timeout_ms
+          = reflection::adl<std::chrono::milliseconds>{}.from(val_reader);
+    }
```

Comment:
> cluster: Add more info into fence batch
Fence batch will contain info about seq_number
for transaction. Becasue we do not wanna use prepare
for rm_stm we need store seq_number on begin

Changes to serialization are very hard to undo / workaround after code ships. Because adl<> doesn't have a lot of the nice features that serde does, we need to communicate clearly what is happening when we make changes to adl-based serialization.

The commit message needs to explain the finer points on changes to versioning in this commit. For example, what happens if an older node reads a message from a newer node? If that's not possible, then explain why.